oteapi/strategies/parse/postgres.py

```diff
@@ -35,78 +33,78 @@ class PostgresResourceConfig(ResourceConfig):
         description="Configuration options for the local data cache.",
     )
 
+    @model_validator(mode="before")
     @classmethod
-    def _urlconstruct(
-        cls,  # PEP8 - Always use cls for the first argument to class methods.
-        scheme: Optional[str] = "",  # Schema defining link format
-        user: Optional[str] = None,  # Username
-        password: Optional[str] = None,  # Password
-        host: Optional[str] = None,
-        port: Optional[int] = None,
-        path: Optional[str] = "",
-        params: Optional[str] = "",
-        query: Optional[str] = "",
-        fragment: Optional[str] = "",
-    ):
-        """Construct a pydantic AnyUrl based on the given URL properties"""
-
-        # Hostname should always be given
-        if not host:
-            raise ValueError("hostname must be specified")
-
-        # Update netloc of username or username|password pair is defined
-        netloc = host
-        if user and not password:  # Only username is provided. OK
-            netloc = f"{user}@{host}"
-        elif user and password:  # Username and password is provided. OK
-            netloc = f"{user}:{password}@{host}"
-        else:  # Password and no username is provided. ERROR
-            raise ValueError("username not provided")
-
-        # Append port if port is defined
-        netloc = netloc if not port else f"{netloc}:{port}"
-
-        # Construct a URL from a tuple of URL-properties
-        unparsed = urlunparse([scheme, netloc, path, params, query, fragment])
-
-        # Populate and return a Pydantic URL
-        return parse_obj_as(AnyUrl, unparsed)
-
-    @root_validator
-    def adjust_url(cls, values):
-        """Root Validator
+    def adjust_url(cls, data: Any) -> "PostgresResourceConfig":
+        """Model Validator
         Verifies configuration consistency, merge configurations
         and update the accessUrl property.
         """
+        if isinstance(data, BaseModel):
+            data = data.model_dump()
+        elif not isinstance(data, dict):
+            raise TypeError(
+                "invalid data type, should be either dict or pydantic model"
+            )
+
+        if "accessUrl" not in data:
+            raise ValueError("missing accessUrl in PostgreSQL resource configuration")
 
-        # Copy model-state into placeholders
-        config = values.get("configuration")
-        accessUrl = values["accessUrl"]
+        accessUrl = AnyUrl(data["accessUrl"])
+        default_config = PostgresConfig()
+
+        if not accessUrl.host:
+            raise ValueError("missing host in accessUrl")
 
         # Check and merge user configuration
-        user = accessUrl.user if accessUrl.user else config["user"]
-        if config["user"] and user != config["user"]:
+        user = (
```

Comment:
the method seems very hard to read and verbose. I'm not very clear why this has been changed, is it for compatibility reasons?

Reply:
I suppose a reversion to using placeholder variables can be done for verbosity's and readability's sake.

Reply:
Done

Reply:
it still seems like we can maybe refactor a bit more, how do you feel about something like: 

```python
if "accessUrl" not in data:
    raise ValueError("missing accessUrl in PostgreSQL resource configuration")

# Copy model-state into placeholders
accessUrl = AnyUrl(data["accessUrl"])
default_config = PostgresConfig()
current_config: dict[str, Any] = data.get("configuration", {})

if not accessUrl.host:
    raise ValueError("missing host in accessUrl")

def _get_and_validate_config_value(attr, config_key):
    """Get value from accessUrl or current_config, and check for mismatches."""
    value_from_url = getattr(accessUrl, attr)
    value_from_config = current_config.get(config_key, getattr(default_config, config_key))
    
    final_value = value_from_url or value_from_config
    
    if value_from_config and final_value != value_from_config:
        raise ValueError(f"mismatching {config_key} in accessUrl and configuration")
    
    return final_value

user = _get_and_validate_config_value("username", "user")
password = _get_and_validate_config_value("password", "password")
dbname = _get_and_validate_config_value("path", "dbname")
```